test/IntegrationTests/MSTest.Acceptance.IntegrationTests/ThreadContextTests.cs

```diff
@@ -117,12 +209,562 @@ public void TestMethod1()
+
+#file CultureFlowsThreadContextProject.csproj
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <EnableMSTestRunner>true</EnableMSTestRunner>
+    <TargetFrameworks>$TargetFrameworks$</TargetFrameworks>
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="MSTest.TestAdapter" Version="$MSTestVersion$" />
+    <PackageReference Include="MSTest.TestFramework" Version="$MSTestVersion$" />
+    <PackageReference Include="Microsoft.Testing.Platform" Version="$MicrosoftTestingPlatformVersion$" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="*.runsettings">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
+</Project>
+
+#file UnitTest1.cs
+namespace CultureFlowsThreadContextProject;
+
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+[TestClass]
+public class UnitTest1
+{
+    private const string AssemblyInitCultureCodeName = "th-TH";
+    private const string ClassInitCultureCodeName = "tr-TR";
+    private const string TestInitCultureCodeName = "sv-SE";
+    private const string TestMethodCultureCodeName = "ak-GH";
+    private const string TestCleanupCultureCodeName = "pt-BR";
+    private const string ClassCleanupCultureCodeName = "hu-HU";
+
+    [AssemblyInitialize]
+    public static void AssemblyInitialize(TestContext context)
+    {
+        CultureInfo.CurrentCulture = new CultureInfo(AssemblyInitCultureCodeName);
+    }
+
+    [ClassInitialize]
+    public static void ClassInitialize(TestContext context)
+    {
+        Assert.AreEqual(AssemblyInitCultureCodeName, CultureInfo.CurrentCulture.Name, "ClassInitialize culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(ClassInitCultureCodeName);
+    }
+
+    [TestInitialize]
+    public void TestInitialize()
+    {
+        Assert.AreEqual(ClassInitCultureCodeName, CultureInfo.CurrentCulture.Name, "TestInitialize culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestInitCultureCodeName);
+    }
+
+    [TestMethod]
+    public void TestMethod1()
+    {
+        Assert.AreEqual(TestInitCultureCodeName, CultureInfo.CurrentCulture.Name, "TestMethod culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestMethodCultureCodeName);
+    }
+
+    [TestCleanup]
+    public void TestCleanup()
+    {
+        Assert.AreEqual(TestMethodCultureCodeName, CultureInfo.CurrentCulture.Name, "TestCleanup culture check failed");
+        CultureInfo.CurrentCulture = new CultureInfo(TestCleanupCultureCodeName);
+    }
+
+    [ClassCleanup]
+    public static void ClassCleanup()
+    {
+        if (Environment.GetEnvironmentVariable("MSTEST_TEST_FLOW_CONTEXT") == "true")
+        {
+            Assert.AreEqual(ClassInitCultureCodeName, CultureInfo.CurrentCulture.Name, "ClassCleanup culture check failed");
```

Comment:
Once the behaviors are complete, it would be helpful if the error was more concrete about the intent. E.g. ClassCleanup should have the same culture as ClassInit, because ....

Reply:
Agreed, I'll refactor once we all agree on the direction.